modules/gapi/src/streaming/onevpl/demux/async_mfp_demux_data_provider.cpp

```diff
@@ -0,0 +1,783 @@
+            // Get the video frame buffer from the sample.
+            IMFMediaBuffer *buffer_ptr = nullptr;
+            hr = sample_ptr->ConvertToContiguousBuffer(&buffer_ptr);
+            if (FAILED(hr)) {
+                GAPI_Assert(SUCCEEDED(hr) && "MFPAsyncDemuxDataProvider::OnReadSample");
+            }
+
+            DWORD max_buffer_size = 0;
+            DWORD curr_size = 0;
+
+            // lock buffer directly into mfx bitstream
+            std::shared_ptr<mfxBitstream> staging_stream = std::make_shared<mfxBitstream>();
+            staging_stream->Data = nullptr;
+
+            hr = buffer_ptr->Lock(&staging_stream->Data, &max_buffer_size, &curr_size);
+            if (FAILED(hr)) {
+                GAPI_Assert(SUCCEEDED(hr) && "MFPAsyncDemuxDataProvider::OnReadSample");
+            }
+
+            staging_stream->MaxLength = max_buffer_size;
+            staging_stream->DataLength = curr_size;
+            staging_stream->CodecId = get_mfx_codec_id();
+
+            GAPI_LOG_DEBUG(nullptr, "[" << this << "] bitstream created, data: " <<
+                                    static_cast<void*>(staging_stream->Data) <<
+                                    ", MaxLength: " << staging_stream->MaxLength <<
+                                    ", DataLength: "  << staging_stream->DataLength);
+
+            worker_buffer_count = produce_worker_data(staging_stream->Data,
+                                                      createCOMPtrGuard(buffer_ptr),
+                                                      std::move(staging_stream));
+        }
+    } else {
+        GAPI_LOG_WARNING(nullptr, "[" << this << "] callback failed"
+                                  ", status: " << std::to_string(HRESULT_CODE(status)) <<
+                                  ", stream flags: " << stream_flag <<
+                                  ", sample: " << sample_ptr);
+    }
+
+    hr = request_next(hr, stream_flag, worker_buffer_count);
+    return hr;
+}
+
+size_t MFPAsyncDemuxDataProvider::get_locked_buffer_size() const {
+    std::unique_lock<std::mutex> l(buffer_storage_mutex);
+    return worker_locked_buffer_storage.size();
+}
+
+STDMETHODIMP MFPAsyncDemuxDataProvider::OnEvent(DWORD, IMFMediaEvent *) {
+    return S_OK;
+}
+
+STDMETHODIMP MFPAsyncDemuxDataProvider::OnFlush(DWORD) {
+    provider_state.store(State::Exhausted);
+    buffer_storage_non_empty_cond.notify_all();
+    return S_OK;
+}
+
+void MFPAsyncDemuxDataProvider::flush() {
+    if(source_reader) {
+        GAPI_LOG_INFO(nullptr, "[" << this << "] set flush");
+        source_reader->Flush(static_cast<DWORD>(MF_SOURCE_READER_ALL_STREAMS));
+    }
+
+    size_t iterations = 1;
+    while (provider_state.load() != State::Exhausted) {
+        GAPI_LOG_DEBUG(nullptr, "[" << this << "] is waiting for flush finishing, "
+                                "iteration: " <<iterations);
+        iterations++;
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
Don't quite get this. Could you explain this part, please?

Reply:
I have  changed unconditional `sleep` for `wait_for` conditional waiting and  looks like all unit tests have passed.

Reply:
Got it, thanks! Anyway, I think it should be much faster than 100ms